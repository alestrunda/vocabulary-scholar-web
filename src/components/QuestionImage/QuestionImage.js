/*global Image*/

import React from 'react';
import PropTypes from 'prop-types';
import Hint from '../Hint';
import LoaderText from '../LoaderText';

class QuestionImage extends React.Component {
  state = {
    isLoaded: false,
    isError: false,
  };

  componentDidMount() {
    this.loadImage(this.props.imageSrc);
  }

  componentDidUpdate(prevProps) {
    if (this.props.wordID !== prevProps.wordID) {
      this.setState({
        isLoaded: false,
        isError: false,
      });
      this.loadImage(this.props.imageSrc);
    }
  }

  handleQuestionReady() {
    this.props.onQuestionReady && this.props.onQuestionReady();
  }

  loadImage(imgSrc) {
    if (!imgSrc) {
      this.setState({
        isError: true,
      });
      return;
    }
    const imageElement = new Image();
    imageElement.onload = () => {
      this.setState({
        isLoaded: true,
      });
      this.handleQuestionReady();
    };
    imageElement.onerror = () => {
      this.setState({
        isError: true,
      });
    };
    imageElement.src = imgSrc;
  }

  render() {
    return (
      <div>
        <div className={this.props.className} style={this.props.style}>
          {!this.state.isLoaded && !this.state.isError && <LoaderText />}
          {this.state.isError && (
            <p className="text-no-data">Image not available</p>
          )}
          {this.state.isLoaded && <img alt="" src={this.props.imageSrc} />}
        </div>
        {this.state.isLoaded && (
          <Hint
            className="mt10"
            hint={this.props.hint}
            hintType={this.props.hintType}
            wordID={this.props.wordID}
          />
        )}
      </div>
    );
  }
}

QuestionImage.propTypes = {
  className: PropTypes.string,
  word: PropTypes.string,
  wordID: PropTypes.string,
  style: PropTypes.object,
  imageSrc: PropTypes.string,
  onQuestionReady: PropTypes.func,
  hint: PropTypes.string,
  hintType: PropTypes.string,
};

export default QuestionImage;
