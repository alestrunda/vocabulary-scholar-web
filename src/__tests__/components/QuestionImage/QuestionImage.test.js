import React from 'react';
import { shallow } from 'enzyme';

import QuestionImage from '../../../components/QuestionImage';

it('renders loading', () => {
  const root = shallow(<QuestionImage imageSrc="file.mp3" wordID="school" />);
  expect(root).toMatchSnapshot();
});

it('renders error', () => {
  const root = shallow(<QuestionImage imageSrc="file.mp3" wordID="school" />);
  root.setState({
    error: true,
  });
  expect(root).toMatchSnapshot();
});

it('renders correctly', () => {
  const root = shallow(<QuestionImage imageSrc="file.mp3" wordID="school" />);
  root.setState({
    loaded: true,
  });
  expect(root).toMatchSnapshot();
});
