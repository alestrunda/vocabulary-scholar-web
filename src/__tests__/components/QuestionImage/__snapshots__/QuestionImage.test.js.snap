// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <QuestionImage
    imageSrc="file.mp3"
    wordID="school"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div>
          <LoaderText
            text="Loading"
          />
        </div>,
        false,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <LoaderText
              text="Loading"
            />,
            false,
            false,
          ],
          "className": undefined,
          "style": undefined,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "text": "Loading",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          false,
          false,
        ],
        "type": "div",
      },
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div>
            <LoaderText
              text="Loading"
            />
          </div>,
          false,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <LoaderText
                text="Loading"
              />,
              false,
              false,
            ],
            "className": undefined,
            "style": undefined,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "text": "Loading",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            false,
            false,
          ],
          "type": "div",
        },
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`renders error 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <QuestionImage
    imageSrc="file.mp3"
    wordID="school"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div>
          <LoaderText
            text="Loading"
          />
        </div>,
        false,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <LoaderText
              text="Loading"
            />,
            false,
            false,
          ],
          "className": undefined,
          "style": undefined,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "text": "Loading",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          false,
          false,
        ],
        "type": "div",
      },
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div>
            <LoaderText
              text="Loading"
            />
          </div>,
          false,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <LoaderText
                text="Loading"
              />,
              false,
              false,
            ],
            "className": undefined,
            "style": undefined,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "text": "Loading",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            false,
            false,
          ],
          "type": "div",
        },
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`renders loading 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <QuestionImage
    imageSrc="file.mp3"
    wordID="school"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div>
          <LoaderText
            text="Loading"
          />
        </div>,
        false,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <LoaderText
              text="Loading"
            />,
            false,
            false,
          ],
          "className": undefined,
          "style": undefined,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "text": "Loading",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          false,
          false,
        ],
        "type": "div",
      },
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div>
            <LoaderText
              text="Loading"
            />
          </div>,
          false,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <LoaderText
                text="Loading"
              />,
              false,
              false,
            ],
            "className": undefined,
            "style": undefined,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "text": "Loading",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            false,
            false,
          ],
          "type": "div",
        },
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
